### PR TITLE
Azure-cosmos-Spark - allow configuration of initial batch size

### PR DESCRIPTION
**Pull request by FabianMeiswinkel**
# Description

## Problem statement
When using the Cosmos DB Spark connector as a sink (writing to Cosmos DB from Spark) the size of a micro batch (how many documents will be ingested within a single request to the service) is automatically tuned based on the throttling rate. This is done to ensure that the provisioned throughput is really fully saturated without causing so much noise due to throttled requests that it effectively reduces the total throughput. When enabling throughput for an ingestion job, the tuning of the micro batch size also happens for requests that are throttled on the client side - when the SDK determines that the RU/s the client has consumed is higher than the `spark.cosmos.throughputControl.targetThroughput` / `spark.cosmos.throughputControl.targetThroughputThreshold`. This works well to ensure the average throughput consumed stays within the limit of the targeted threshold plus/minus a few percent. But it can still result in exceeding the targeted throughput especially in the first few seconds of the job - when the batch size gets automatically tuned. The default initial batch size is 100 - let's assume the throughput allows using a batch size of 3 only - in the initialization phase of the job, the micro batch size based on the throttling rate gets reduce from the default `100` eventually to `3` - but while this tuning is happening it will consume more RU/s than the targeted throughput specified. 

## Solution
Usually this won't be a problem - throttled requests are automatically retried, and tuning the batch size down will happen within a few seconds. In general there is no need to change this behavior. But if for a specific workload this is a problem, the new configuration setting introduced in this PR `spark.cosmos.write.bulk.initialBatchSize` allows specifying the initial batch size - in most cases you will just want to start with setting `spark.cosmos.write.bulk.initialBatchSize` to `1` and let the automatic tuning increase the batch size over time.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala [hunk 1]
```diff
@@ -94,6 +94,28 @@ class BulkWriter(container: CosmosAsyncContainer,
...
+          customMaxMicroBatchPayloadSizeInBytes
+        )
+    case None =>
+  }
+
+  writeConfig.initialMicroBatchSize match {
+    case Some(customInitialMicroBatchSize) =>
+      ImplementationBridgeHelpers.CosmosBulkExecutionOptionsHelper
+        .getCosmosBulkExecutionOptionsAccessor
+        .setMaxMicroBatchSize(
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala:112]: do we need to introduce a new parameter for initialMicroBatchSize? - the partitionScopeThreshold is bounded by maxBatchSize as well, so i do not think we are gonna able to tune back to 100

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/BulkWriter.scala]: Done

File: sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/linkedin/DataLoader.java [hunk 2]
```diff
@@ -94,10 +94,7 @@ private void bulkCreateItems(final CosmosAsyncDatabase database,
             container.getId());
 
         // We want to wait longer depending on the number of documents in each iteration
-        final CosmosBulkExecutionOptions cosmosBulkExecutionOptions =
-            ImplementationBridgeHelpers.CosmosBulkExecutionOptionsHelper
-                .getCosmosBulkExecutionOptionsAccessor()
-                .setMaxMicroBatchSize(new CosmosBulkExecutionOptions(), MAX_BATCH_SIZE);
+        final CosmosBulkExecutionOptions cosmosBulkExecutionOptions =new CosmosBulkExecutionOptions();
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-benchmark/src/main/java/com/azure/cosmos/benchmark/linkedin/DataLoader.java:97]: nit: space.` =new `

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosBulkExecutionOptions.java [hunk 3]
```diff
@@ -70,41 +70,12 @@ public CosmosBulkExecutionOptions() {
...
-     * documents are of size 20KB, approximately 90 operations will always be retried. So it's better to choose a batch
-     * size of 10 here if user is aware of there workload. If sizes are totally unknown and user cannot put a number on it
-     * then retries are handled, so no issues as such.
-     *
-     * If the retry rate exceeds `getMaxMicroBatchInterval` the micro batch size gets dynamically reduced at runtime
-     * @return micro batch size
-     */
-    int getMaxMicroBatchSize() {
-        return maxMicroBatchSize;
+    int getInitialMicroBatchSize() {
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosBulkExecutionOptions.java:73]: maybe add some java doc for this property?

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala [hunk 4]
```diff
@@ -839,6 +842,16 @@ private object CosmosWriteConfig {
       "room for one document - to avoid that the request size exceeds the Cosmos DB maximum of 2 MB too often " +
       "which would result in retries and having to transmit large network payloads multiple times.")
 
+  private val initialMicroBatchSize = CosmosConfigEntry[Int](key = CosmosConfigNames.WriteBulkInitialBatchSize,
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:845]: do we need to public document it in the configuration-reference?

**Review by xinlian12 — APPROVED** [review]
LGTM, thanks :)

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosBulkExecutionOptions.java (hunk 3 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/models/CosmosBulkExecutionOptions.java:73]: It is intenral - intentionally not added - what else than the name already tells?

File: sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala (hunk 4 above)
Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:845]: Yes - will do that in separate PR

Comment (FabianMeiswinkel) [reply sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosConfig.scala:845]: Good catch

**Pull request merged**